### PR TITLE
migrate `thumb-none-cortex-m` to rmake

### PR DESCRIPTION
**Pull request by folkertdev**
tracking issue: https://github.com/rust-lang/rust/issues/121876

I'll leave some comments/questions inline

r? @jieyouxu

try-job: armhf-gnu
try-job: dist-various-1
try-job: test-various

### CODE REVIEW

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (folkertdev) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:20]: this version is pretty old. That has advantages (should still work), and I didn't want to change it now so that this test still tests the same thing. Nonetheless, maybe this should be updated at some point?

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (folkertdev) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:27]: this is what the makefile did, but it seems to be unique among these tests. We could download the artifacts for this specific release directly too, but maybe git is easier/just always installed anyway?

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (folkertdev) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:47]: should this be extracted into some `cargo` helper? it occurs a couple of times and is kind of inscrutable with all the env variables.

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (folkertdev) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:55]: is this required for these embedded targets? I guess it was in the makefile for a reason?

File: tests/run-make/thumb-none-cortex-m/Makefile
Comment (folkertdev) [inline tests/run-make/thumb-none-cortex-m/Makefile:37]: is this comment still relevant?

Comment (folkertdev) [reply tests/run-make/thumb-none-cortex-m/Makefile:37]: setting that environment variable is not relevant on my system anyway. It's still in the code right now to be a faithful translation, but maybe this can be dropped now?

**Comment by jieyouxu** [comment]
Oof. Thank you for taking this one on, this is one of the tests I spoke with @Oneirical about being very ~~cursed~~ *interesting*. Expect some delay from me, because I probably will also be asking around for help about what this is trying to test and in particular how it achieves that 😅

**Comment by folkertdev** [comment]
yeah I figured that this one was a bit of an outlier. I and several of my colleagues are quite familiar with this target platform, so we're happy to help out where that expertise is required.

**Comment by jieyouxu** [comment]
Update: let me get back to you after discussing this test (`thumb-none-cortex-m`) and `thumb-none-qemu` with the compiler people regarding needing internet connection (due to cargo external deps) in order to run the test...

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (jieyouxu) [reply tests/run-make/thumb-none-cortex-m/rmake.rs:20]: Personally, I would be fine bumping this to a newer version.

File: tests/run-make/thumb-none-cortex-m/Makefile
Comment (jieyouxu) [reply tests/run-make/thumb-none-cortex-m/Makefile:37]: I believe `RUSTC_BOOTSTRAP=1` is not needed if the crate and its transitive dependencies do not depend on nightly features.

**Comment by jieyouxu** [comment]
I don't want to block this PR on the external dep challenges, they can be done separately (cc #128733).
@rustbot review

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (jieyouxu) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:1]: Suggestion: we should explain what this test is trying to check. Maybe something to the effect of "a basic smoke test exercising cargo + rustc to see if the cortex-m crate can be compiled successfully for thumb targets".

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (jieyouxu) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:9]: Question: do we know if these instructions and support targets are up-to-date?

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (jieyouxu) [inline tests/run-make/thumb-none-cortex-m/rmake.rs:30]: Suggestion: `target()` helper from run_make_support.

**Review by jieyouxu — APPROVED** [review]
Thank you for tackling this, I have some suggestions, but this LGTM in general.

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (jieyouxu) [reply tests/run-make/thumb-none-cortex-m/rmake.rs:30]: @rustbot author

**Comment by jieyouxu** [comment]
In the meantime, let's see what our CI try jobs say.
@bors try

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (folkertdev) [reply tests/run-make/thumb-none-cortex-m/rmake.rs:13]: based on https://hackmd.io/@thejpster/H1iTNJSRT the comment can be extended by

//! - thumbv8m.base-none-eabi (Bare Cortex-M23)
//! - thumbv8m.main-none-eabi (Bare Cortex-M33, M55, M85)
//! - thumbv8m.main-none-eabihf (Bare Cortex-M33, M55, M85, FPU, hardfloat)

this appears to work fine locally, but I think is only useful if CI actually runs these targets

**Review by folkertdev — COMMENTED** [review]
@rustbot ready

I can squash the commits if that's better, but know that github's interface does not always work well with force pushes.

we might want to try this for a windows host (due to removing the LIB variable)

File: tests/run-make/thumb-none-cortex-m/rmake.rs
Comment (jieyouxu) [reply tests/run-make/thumb-none-cortex-m/rmake.rs:13]: > it is correct, but today probably should also include
> 
>     * thumbv8m.base-none-eabi
>     * thumbv8m.main-none-eabi
>     * thumbv8m.main-none-eabihf
> 
> 
> I'm not sure whether CI even runs those though?

cc @rust-lang/infra dear infra people, do we know if any of our CI jobs exercise these targets?

**Review by jieyouxu — APPROVED** [review]
Thanks, this looks good to me. I'll run try-jobs to check.

**Pull request merged**
